### ISSUE TITLE
GitHub Actions CI troubleshooting

### ISSUE THREAD

**Issue by ExeRhythm** [SignTools/ios-signer-service#7]
Hi again. I have another issue where GitHub Actions cannot sign my .iPA file. It says 
```
/Users/runner/work/_temp/39df6836-3324-44f5-8d3f-c6d7d3fbe509.sh: line 1: ./sign.sh: No such file or directory
Error: Process completed with exit code 1."
```
Here's the screenshot:
<img width="1076" alt="Screenshot 2021-03-03 at 21 17 17" src="https://user-images.githubusercontent.com/52459150/109852479-ddd1fe00-7c65-11eb-80ed-afa901e1d42f.png">
Also, sorry for asking too much help. I love your project.

**Comment by ViRb3** [comment]
Oh, I know exactly what this is, an oversight after I made some changes. I will fix it in a second and let you know.

**Comment by ViRb3** [comment]
I pushed the fix, please update your fork and try again. Let me know how it goes.

**Comment by ExeRhythm** [comment]
Thanks, it worked. But now I have another similar issue: 
```
Run ./sign.sh
Obtaining files...
cat: pass.txt: No such file or directory
Error: Process completed with exit code 1.
```
Did I do something wrong?

**Comment by ViRb3** [comment]
This means that the builder could not properly download a job from your server. I pushed another update to the CI repo that will print the error instead of swallowing it. Please update, try again, and share the error here.

**Comment by ExeRhythm** [comment]
Here is the log.
<img width="372" alt="Screenshot 2021-03-03 at 21 48 55" src="https://user-images.githubusercontent.com/52459150/109856191-43c08480-7c6a-11eb-8451-367810992a4d.png">

**Comment by ViRb3** [comment]
This looks like a different error from the one above. Here it fails during the signing process. To debug this, go into the `sign.sh` file in your CI repo and just below the `Signing...` line, change:
```bash
./xresign.sh -i unsigned.ipa -c "$IDENTITY" -p "prov.mobileprovision" $SIGN_ARGS > /dev/null 2>&1
```
to:
```bash
./xresign.sh -i unsigned.ipa -c "$IDENTITY" -p "prov.mobileprovision" $SIGN_ARGS
```
(remove the output suppression)

I added that so it doesn't leak potentially private information about the app you're signing. Feel free to share the error here if you can't figure it out.

**Comment by ExeRhythm** [comment]
<img width="478" alt="Screenshot 2021-03-03 at 21 59 40" src="https://user-images.githubusercontent.com/52459150/109857458-c3028800-7c6b-11eb-8364-87ada8c08b53.png">

**Comment by ViRb3** [comment]
Very weird. From https://github.com/SignTools/XReSign/blob/master/XReSign/Scripts/xresign.sh:
```bash
if [ -z "$SOURCEIPA" ] || [ -z "$DEVELOPER" ]; then
    echo "$usage" >&2
    exit 1
fi
```
This is the error that it's hitting. It means that either the "-c" or "-i" argument is empty. How are you running the workflow? It's being triggered by the service when you upload an app, correct?

**Comment by ExeRhythm** [comment]
Right

**Comment by ViRb3** [comment]
So "-i" can't be empty as we're setting it to "unsigned.ipa". This only leaves "-c" being empty. That comes from here in `sign.sh`:
```bash
IDENTITY=$(security find-identity -p appleID -v |  grep -o '".*"' | cut -d '"' -f 2)
```
I suppose this command doesn't successfully obtain the certificate common name. Maybe the certificate isn't being properly imported? 

In this line:
```bash
security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k "1234" > /dev/null 2>&1
```
can you remove the output suppression bit again (`> /dev/null 2>&1`), and see what it says?
Could you also add a line `echo $IDENTITY` there, just to verify this?

**Comment by ExeRhythm** [comment]
Sure, give me just a second 👍🏻

**Comment by ExeRhythm** [comment]
```
Obtaining files...
8
Creating keychain...
9
Importing certificate...
10
keychain: "/Users/runner/Library/Keychains/sign-db"
11
version: 512
12
class: 0x00000010 
13
attributes:
14
    0x00000000 <uint32>=0x00000010 
15
    0x00000001 <blob>="Apple Development: Marina Anisovich (Marina Anisovich)"
16
    0x00000002 <blob>=<NULL>
17
    0x00000003 <uint32>=0x00000001 
18
    0x00000004 <uint32>=0x00000000 
19
    0x00000005 <uint32>=0x00000000 
20
    0x00000006 <blob>=0x3E6A47DB030B3754EFBA6CFFA52A6AE8ADE3F950  ">jG\333\003\0137T\357\272l\377\245*j\350\255\343\371P"
21
    0x00000007 <blob>=<NULL>
22
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
23
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
24
    0x0000000A <uint32>=0x00000800 
25
    0x0000000B <uint32>=0x00000800 
26
    0x0000000C <blob>=0x0000000000000000 
27
    0x0000000D <blob>=0x0000000000000000 
28
    0x0000000E <uint32>=0x00000001 
29
    0x0000000F <uint32>=0x00000001 
30
    0x00000010 <uint32>=0x00000001 
31
    0x00000011 <uint32>=0x00000000 
32
    0x00000012 <uint32>=0x00000001 
33
    0x00000013 <uint32>=0x00000001 
34
    0x00000014 <uint32>=0x00000001 
35
    0x00000015 <uint32>=0x00000001 
36
    0x00000016 <uint32>=0x00000001 
37
    0x00000017 <uint32>=0x00000001 
38
    0x00000018 <uint32>=0x00000001 
39
    0x00000019 <uint32>=0x00000001 
40
    0x0000001A <uint32>=0x00000001 
41
keychain: "/Library/Keychains/System.keychain"
42
version: 256
43
class: 0x00000010 
44
attributes:
45
    0x00000000 <uint32>=0x00000010 
46
    0x00000001 <blob>="com.apple.systemdefault"
47
    0x00000002 <blob>=<NULL>
48
    0x00000003 <uint32>=0x00000001 
49
    0x00000004 <uint32>=0x00000000 
50
    0x00000005 <uint32>=0x00000000 
51
    0x00000006 <blob>=0x2707CC3BC956330A0B08D3733F816A78CCCEF24A  "'\007\314;\311V3\012\013\010\323s?\201jx\314\316\362J"
52
    0x00000007 <blob>=<NULL>
53
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
54
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
55
    0x0000000A <uint32>=0x00000800 
56
    0x0000000B <uint32>=0x00000800 
57
    0x0000000C <blob>=0x0000000000000000 
58
    0x0000000D <blob>=0x0000000000000000 
59
    0x0000000E <uint32>=0x00000001 
60
    0x0000000F <uint32>=0x00000001 
61
    0x00000010 <uint32>=0x00000001 
62
    0x00000011 <uint32>=0x00000000 
63
    0x00000012 <uint32>=0x00000000 
64
    0x00000013 <uint32>=0x00000001 
65
    0x00000014 <uint32>=0x00000000 
66
    0x00000015 <uint32>=0x00000001 
67
    0x00000016 <uint32>=0x00000000 
68
    0x00000017 <uint32>=0x00000000 
69
    0x00000018 <uint32>=0x00000000 
70
    0x00000019 <uint32>=0x00000000 
71
    0x0000001A <uint32>=0x00000001 
72
keychain: "/Library/Keychains/System.keychain"
73
version: 256
74
class: 0x00000010 
75
attributes:
76
    0x00000000 <uint32>=0x00000010 
77
    0x00000001 <blob>="com.apple.kerberos.kdc"
78
    0x00000002 <blob>=<NULL>
79
    0x00000003 <uint32>=0x00000001 
80
    0x00000004 <uint32>=0x00000000 
81
    0x00000005 <uint32>=0x00000000 
82
    0x00000006 <blob>=0x7EEBE50F24D2825F959079443DF02DD698A0A50B  "~\353\345\017$\322\202_\225\220yD=\360-\326\230\240\245\013"
83
    0x00000007 <blob>=<NULL>
84
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
85
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
86
    0x0000000A <uint32>=0x00000800 
87
    0x0000000B <uint32>=0x00000800 
88
    0x0000000C <blob>=0x0000000000000000 
89
    0x0000000D <blob>=0x0000000000000000 
90
    0x0000000E <uint32>=0x00000001 
91
    0x0000000F <uint32>=0x00000001 
92
    0x00000010 <uint32>=0x00000001 
93
    0x00000011 <uint32>=0x00000000 
94
    0x00000012 <uint32>=0x00000000 
95
    0x00000013 <uint32>=0x00000001 
96
    0x00000014 <uint32>=0x00000000 
97
    0x00000015 <uint32>=0x00000001 
98
    0x00000016 <uint32>=0x00000000 
99
    0x00000017 <uint32>=0x00000000 
100
    0x00000018 <uint32>=0x00000000 
101
    0x00000019 <uint32>=0x00000000 
102
    0x0000001A <uint32>=0x00000001 
103
1.2.840.113635.100.1.61
104
Signing...
105
Usage: xresign.sh -i APP_PATH -c CERT_NAME [-epbdas ...]
106

107
-i  path to input app to sign
108
-c  Common Name of signing certificate in Keychain
109
-e  new entitlements to use for app (Optional)
110
-p  path to mobile provisioning file (Optional)
111
-b  new bundle identifier (Optional)
112
-d  enable app debugging (get-task-allow) (Optional)
113
-a  force enable support for all devices (Optional)
114
-s  force enable file sharing through iTunes (Optional)
115
Error: Process completed with exit code 1.
```

**Comment by ViRb3** [comment]
Am I right to understand that `echo $IDENTITY` printed nothing, so the common name reading failed?

**Comment by ExeRhythm** [comment]
Oh wait, forgot to add it. Will upload another app right now...

**Comment by ExeRhythm** [comment]
`Run ./sign.sh
2
  ./sign.sh
3
  shell: /bin/bash -e {0}
4
  env:
5
    SECRET_URL: ***
6
    SECRET_KEY: ***
8
Obtaining files...
9
Creating keychain...
10
Importing certificate...
11
keychain: "/Users/runner/Library/Keychains/sign-db"
12
version: 512
13
class: 0x00000010 
14
attributes:
15
    0x00000000 <uint32>=0x00000010 
16
    0x00000001 <blob>="Apple Development: Marina Anisovich (Marina Anisovich)"
17
    0x00000002 <blob>=<NULL>
18
    0x00000003 <uint32>=0x00000001 
19
    0x00000004 <uint32>=0x00000000 
20
    0x00000005 <uint32>=0x00000000 
21
    0x00000006 <blob>=0x3E6A47DB030B3754EFBA6CFFA52A6AE8ADE3F950  ">jG\333\003\0137T\357\272l\377\245*j\350\255\343\371P"
22
    0x00000007 <blob>=<NULL>
23
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
24
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
25
    0x0000000A <uint32>=0x00000800 
26
    0x0000000B <uint32>=0x00000800 
27
    0x0000000C <blob>=0x0000000000000000 
28
    0x0000000D <blob>=0x0000000000000000 
29
    0x0000000E <uint32>=0x00000001 
30
    0x0000000F <uint32>=0x00000001 
31
    0x00000010 <uint32>=0x00000001 
32
    0x00000011 <uint32>=0x00000000 
33
    0x00000012 <uint32>=0x00000001 
34
    0x00000013 <uint32>=0x00000001 
35
    0x00000014 <uint32>=0x00000001 
36
    0x00000015 <uint32>=0x00000001 
37
    0x00000016 <uint32>=0x00000001 
38
    0x00000017 <uint32>=0x00000001 
39
    0x00000018 <uint32>=0x00000001 
40
    0x00000019 <uint32>=0x00000001 
41
    0x0000001A <uint32>=0x00000001 
42
keychain: "/Library/Keychains/System.keychain"
43
version: 256
44
class: 0x00000010 
45
attributes:
46
    0x00000000 <uint32>=0x00000010 
47
    0x00000001 <blob>="com.apple.systemdefault"
48
    0x00000002 <blob>=<NULL>
49
    0x00000003 <uint32>=0x00000001 
50
    0x00000004 <uint32>=0x00000000 
51
    0x00000005 <uint32>=0x00000000 
52
    0x00000006 <blob>=0x2707CC3BC956330A0B08D3733F816A78CCCEF24A  "'\007\314;\311V3\012\013\010\323s?\201jx\314\316\362J"
53
    0x00000007 <blob>=<NULL>
54
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
55
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
56
    0x0000000A <uint32>=0x00000800 
57
    0x0000000B <uint32>=0x00000800 
58
    0x0000000C <blob>=0x0000000000000000 
59
    0x0000000D <blob>=0x0000000000000000 
60
    0x0000000E <uint32>=0x00000001 
61
    0x0000000F <uint32>=0x00000001 
62
    0x00000010 <uint32>=0x00000001 
63
    0x00000011 <uint32>=0x00000000 
64
    0x00000012 <uint32>=0x00000000 
65
    0x00000013 <uint32>=0x00000001 
66
    0x00000014 <uint32>=0x00000000 
67
    0x00000015 <uint32>=0x00000001 
68
    0x00000016 <uint32>=0x00000000 
69
    0x00000017 <uint32>=0x00000000 
70
    0x00000018 <uint32>=0x00000000 
71
    0x00000019 <uint32>=0x00000000 
72
    0x0000001A <uint32>=0x00000001 
73
keychain: "/Library/Keychains/System.keychain"
74
version: 256
75
class: 0x00000010 
76
attributes:
77
    0x00000000 <uint32>=0x00000010 
78
    0x00000001 <blob>="com.apple.kerberos.kdc"
79
    0x00000002 <blob>=<NULL>
80
    0x00000003 <uint32>=0x00000001 
81
    0x00000004 <uint32>=0x00000000 
82
    0x00000005 <uint32>=0x00000000 
83
    0x00000006 <blob>=0x7EEBE50F24D2825F959079443DF02DD698A0A50B  "~\353\345\017$\322\202_\225\220yD=\360-\326\230\240\245\013"
84
    0x00000007 <blob>=<NULL>
85
    0x00000008 <blob>=0x7B38373139316361322D306663392D313164342D383439612D3030303530326235323132327D00  "{87191ca2-0fc9-11d4-849a-000502b52122}\000"
86
    0x00000009 <uint32>=0x0000002A  "\000\000\000*"
87
    0x0000000A <uint32>=0x00000800 
88
    0x0000000B <uint32>=0x00000800 
89
    0x0000000C <blob>=0x0000000000000000 
90
    0x0000000D <blob>=0x0000000000000000 
91
    0x0000000E <uint32>=0x00000001 
92
    0x0000000F <uint32>=0x00000001 
93
    0x00000010 <uint32>=0x00000001 
94
    0x00000011 <uint32>=0x00000000 
95
    0x00000012 <uint32>=0x00000000 
96
    0x00000013 <uint32>=0x00000001 
97
    0x00000014 <uint32>=0x00000000 
98
    0x00000015 <uint32>=0x00000001 
99
    0x00000016 <uint32>=0x00000000 
100
    0x00000017 <uint32>=0x00000000 
101
    0x00000018 <uint32>=0x00000000 
102
    0x00000019 <uint32>=0x00000000 
103
    0x0000001A <uint32>=0x00000001 
104
1.2.840.113635.100.1.61
105

106
Signing...
107
Usage: xresign.sh -i APP_PATH -c CERT_NAME [-epbdas ...]
108

109
-i  path to input app to sign
110
-c  Common Name of signing certificate in Keychain
111
-e  new entitlements to use for app (Optional)
112
-p  path to mobile provisioning file (Optional)
113
-b  new bundle identifier (Optional)
114
-d  enable app debugging (get-task-allow) (Optional)
115
-a  force enable support for all devices (Optional)
116
-s  force enable file sharing through iTunes (Optional)
117
Error: Process completed with exit code 1.`

**Comment by ViRb3** [comment]
I deleted your logs to preserve any private information that could have been in them.

I suppose the print statement was in line 105, which is empty. That confirms my theory. The question now is, did you export an incorrect certificate, or is my command not working correctly? This is the line that fails:
```bash
IDENTITY=$(security find-identity -p appleID -v |  grep -o '".*"' | cut -d '"' -f 2)
```
Do you have any ideas?

**Comment by ExeRhythm** [comment]
Thank you for the reply! Yes, it returns. Here's the output:
```
1.2.8*******
  1) ******* "Apple Development: ******* (*******QBY)" (Missing required extension)
  2) ******* "Apple Distribution: ******* (*******7W2)" (Missing required extension)
  3) ******* "localhost" (CSSMERR_TP_NOT_TRUSTED)
  4) *******04ABC "Developer ID Application: ******* (*******X27W2)" (Missing required extension)
  5) *******F90DD0ECB "iPhone Developer: ******* (MN6F3NJQBY)" (Missing required extension)
     5 valid identities found

```
(Relaced some parts with *******)

**Comment by ViRb3** [comment]
Can you run the same command on the CI? Meanwhile I will see what possible commands for `-p` there are, and we can see which one returns your certificate.

**Comment by ExeRhythm** [comment]
Also, is this the right certificate I need to export?
<img width="273" alt="Screenshot 2021-03-03 at 22 31 15" src="https://user-images.githubusercontent.com/52459150/109861266-2c849580-7c70-11eb-889c-131ed2d4ec92.png">

**Comment by ViRb3** [comment]
You need the one in the middle, not the key icon, but the certificate icon.

**Comment by ExeRhythm** [comment]
But when I try to export the certificate, Keychain does not allow me to export it in .p12
<img width="316" alt="Screenshot 2021-03-03 at 22 49 31" src="https://user-images.githubusercontent.com/52459150/109863468-b9305300-7c72-11eb-9a52-2bd134745c9c.png">

**Comment by ViRb3** [comment]
Check the edit above. Do you have a Discord or Telegram where I could contact you to troubleshoot this faster?

**Comment by ExeRhythm** [comment]
Yes, @ExeRhythm on telegram and ExeRhythm#6157 on discord

**Comment by ViRb3** [comment]
I deleted your comment again to preserve your account names.

As we discussed over chat, the problem was the way the certificate was exported. You want to select *both* the signing key *and* the certificate, and export them as one single .p12 file. Failing this you end up with incomplete information to achieve signing. I will reflect this in the documentation. Either way, as we verified over chat, you got the CI to work, so I am closing this now.
